helper/resource/teststep_providers_test.go

```diff
@@ -1574,6 +1575,158 @@ func TestTest_TestStep_ProviderFactories_Import_External_WithoutPersistNonMatch(
+								},
+							},
+						},
+					},
+				}, nil
+			},
+		},
+		Steps: []TestStep{
+			{
+				Config: `resource "random_password" "test" { }`,
+				Check:  TestCheckResourceAttr("random_password.test", "min_special", "10"),
+			},
+			{
+				Config:       `resource "random_password" "test" { }`,
+				RefreshState: true,
+				Check:        TestCheckResourceAttr("random_password.test", "min_special", "2"),
+			},
+			{
+				Config: `resource "random_password" "test" { }`,
+				Check:  TestCheckResourceAttr("random_password.test", "min_special", "2"),
+			},
+		},
+	})
+}
+
+func TestTest_TestStep_ProviderFactories_RefreshWithPlanModifier_Inline(t *testing.T) {
+	t.Parallel()
+
+	Test(t, TestCase{
+		ProviderFactories: map[string]func() (*schema.Provider, error){
+			"random": func() (*schema.Provider, error) { //nolint:unparam // required signature
+				return &schema.Provider{
+					ResourcesMap: map[string]*schema.Resource{
+						"random_password": {
+							CustomizeDiff: customdiff.All(
+								func(ctx context.Context, d *schema.ResourceDiff, meta interface{}) error {
+									special := d.Get("special").(bool)
+									if special == true {
+										err := d.SetNew("special", false)
+										if err != nil {
+											panic(err)
+										}
+									}
+									return nil
+								},
+							),
+							CreateContext: func(ctx context.Context, d *schema.ResourceData, i interface{}) diag.Diagnostics {
+								d.SetId("id")
+								err := d.Set("special", false)
+								if err != nil {
+									panic(err)
+								}
+								return nil
+							},
+							DeleteContext: func(_ context.Context, _ *schema.ResourceData, _ interface{}) diag.Diagnostics {
+								return nil
+							},
+							ReadContext: func(_ context.Context, d *schema.ResourceData, _ interface{}) diag.Diagnostics {
+								t := getTimeForTest()
+								if t.After(time.Now().Add(time.Hour * 1)) {
+									err := d.Set("special", true)
+									if err != nil {
+										panic(err)
+									}
+								}
+								return nil
+							},
+							Schema: map[string]*schema.Schema{
+								"special": {
+									Computed: true,
+									Type:     schema.TypeBool,
+									ForceNew: true,
+								},
+
+								"id": {
+									Computed: true,
+									Type:     schema.TypeString,
+								},
+							},
+						},
+					},
+				}, nil
+			},
+		},
+		Steps: []TestStep{
+			{
+				Config: `resource "random_password" "test" { }`,
+				Check:  TestCheckResourceAttr("random_password.test", "special", "false"),
+			},
+			{
+				PreConfig:          setTimeForTest(time.Now().Add(time.Hour * 2)),
+				RefreshState:       true,
+				ExpectNonEmptyPlan: true,
+				Check:              TestCheckResourceAttr("random_password.test", "special", "true"),
+			},
+			{
+				Config:             `resource "random_password" "test" { }`,
+				Check:              TestCheckResourceAttr("random_password.test", "special", "false"),
+				ExpectNonEmptyPlan: true,
```

Comment:
Nit: It'd be nice if the provider resource actually converged after the apply without an unexpected plan difference so it didn't look like this sort of post-RefreshState ExpectNonEmptyPlan was required, but no worries if this is difficult to work out. I think we may be able to reset the time during the `Create` function by explicitly calling something like:

```go
// Ensure test resource can converge after apply.
setTimeForTest(time.Now())
return nil
```

But I'm not sure if that's the least confusing option there.

Reply:
Good point. Have used `PreConfig` to reset time for `TestStep` to avoid state mutation in `ReadContext` and subsequent diff generation by `CustomizeDiff`.